R/write.R

```diff
@@ -0,0 +1,481 @@
+  }
+
+  # Create wb objects
+  wb$add_worksheet(sheet = "Table of Contents")
+  wb$add_data_table(sheet = "Table of Contents",
+                    x = supertbl_toc,
+                    startRow = ifelse(labelled, 2, 1),
+                    tableStyle = table_style)
+  wb$set_col_widths(sheet = "Table of Contents",
+                    cols = seq_along(supertbl_toc),
+                    widths = set_col_widths)
+
+  # Return TOC object as dataframe
+  supertbl_toc
+}
+
+#' @title Add the metadata sheet
+#'
+#' @description
+#' Internal helper function. Adds appropriate elements to wb object. Returns
+#' a dataframe.
+#'
+#' @param supertbl a supertibble generated using `read_redcap()`
+#' @param superbl_meta an `unnest`-ed metadata tibble from the supertibble
+#' @param wb An `openxlsx2` workbook object
+#' @param labelled Whether or not to include labelled outputs, default `FALSE`.
+#' Requires use of `make_labelled()`.
+#' @param include_metadata Include a sheet capturing the combined output of the
+#' supertibble `redcap_metadata`. Default `TRUE`.
+#' @param table_style Any excel table style name or "none" (see "formatting"
+#' vignette in \link[openxlsx2]{wb_add_data_table}). Default "tableStyleLight8".
+#' @param set_col_widths Width to set columns across the workbook. Default
+#' "auto", otherwise a numeric value. Standard Excel is 8.43.
+#'
+#' @importFrom dplyr select filter
+#' @importFrom tidyr unnest
+#'
+#' @returns A dataframe
+#'
+#' @keywords internal
+
+add_metadata_sheet <- function(supertbl,
+                               supertbl_meta,
+                               include_metadata,
+                               wb,
+                               labelled,
+                               table_style,
+                               set_col_widths) {
+
+  wb$add_worksheet(sheet = "REDCap Metadata")
+  wb$add_data_table(sheet = "REDCap Metadata",
+                    x = supertbl_meta,
+                    startRow = ifelse(labelled, 2, 1),
+                    tableStyle = table_style)
+  wb$set_col_widths(sheet = "REDCap Metadata",
+                    cols = seq_along(supertbl_meta),
+                    widths = set_col_widths)
+}
+
+#' @title Check if labelled
+#'
+#' @description
+#' Checks if a supplied supertibble is labelled and throws an error if it is not
+#' but `labelled` is set to `TRUE`
+#'
+#' @param supertbl a supertibble generated using `read_redcap()`
+#' @param labelled Whether or not to include labelled outputs, default `FALSE`.
+#' Requires use of `make_labelled()`.
+#' @param call the calling environment to use in the warning message
+#'
+#' @importFrom cli cli_abort
+#' @importFrom rlang caller_env
+#'
+#' @returns A boolean
+#'
+#' @keywords internal
+
+check_labelled <- function(supertbl, labelled, call = caller_env()) {
+  # supertbl is considered labeleld if any look_for labels not NA
+  label_vec <- supertbl %>%
+    labelled::look_for() %>%
```

Comment:
This needs to be behind a `check_installed()`

Reply:
I still don't think this quite works because a user may hit this call to `labelled::look_for()` without having `labelled` installed. Go to Italy though and we'll figure it out in your absence!

Reply:
Hm yea you're right, I would like to get the installation check to be based on actual user usage and not what we're using for identification. Do you know if labelled tacks on any identifiable attributes that we can access with base R?

Reply:
In case it's helpful, here's a base R function that checks if a dataframe has at least one labelled column:

```r
# Function to check if any column in a data frame is labeled
check_labeled_columns <- function(data) {
    labeled_columns <- sapply(data, function(column) {
        !is.null(attributes(column)$label)
    })
    any(labeled_columns)
}

Reply:
Ha I think I just wrote an almost identical function with `purrr`. One sec